docs/design/0000-00-01-architecture-of-storage-engine-pagestorage.md

```diff
@@ -0,0 +1,237 @@
+# Architecture Of Storage Engine - PageStorage
+
+- Authors(order by last name): [JaySon-Huang](https://github.com/JaySon-Huang), [flowbehappy](https://github.com/flowbehappy), [Jiaqi Zhou](https://github.com/jiaqizho)
+- Technical Writer: [shichun-0415](https://github.com/shichun-0415) , [Jiaqi Zhou](https://github.com/jiaqizho)
+
+## Introduction
+
+`PageStorage` is the place where DT(DeltaTree Engine) actually stores data into. The latest data (i.e. delta data), and the metadata in the engine are serialized directly into Pages. While the main data (i.e. stable data), is written in DTFiles format and managed as `ExternalPage`s in PageStorage.
+
+The below picture describes the "Delta ValueSpace" and "Stable ValueSpace" of DT. The data in "Delta ValueSpace" is continuously updated. After `delta merge` happened, the delta data in PageStorage will be read and compacted into the "Stable ValueSpace".
+
+![tiflash-dt-architecture](./images/tiflash-dt-architecture.png)
+
+
+As one of the important components of DT, PageStorage mainly provides a KV storage service which also support MVCC. Unlike other KV services, the KV interface provided by PageStorage is limited. Key is limited to uint64_t, Value is limited to a buffer or a array of buffer(we called it fields) or null.
+
+
+## Capability
+
+PageStorage supported: 
+
+- Disk-based store
+- Write/Read operation atomicity
+- Full MVCC function 
+- KV store function
+- GC
+
+
+## Design
+
+Currently, there are three versions of PageStorage (V1, V2, V3). We won't describe the details of the V1/V2 in this article. The V2 design and implementation lead to high write amplification and CPU usage under some scenarios, so we propose the V3 version.
+
+### V3 version
+
+After our customers used TiFlash, we found some problems with the V2 version in the actual production environment.
+
+1. **There is a risk of data loss in the meta part**. As long as the `checksum` and `buffer size` fields in a single meta buffer are damaged at the same time, the subsequent buffers will be unavailable.This situation may happen when When the disk failure or the meta been changed due to wrong operation.
+2. **The snapshot of MVCC needs to be optimized**. First, the memory occupied by the snapshot can be reduced. Secondly, There are no need such a complex structure to implement MVCC.
+3. **The GC write amplification in the data part is too severe, The GC task is too heavy**. Since the data part is composed of append write, `compact gc` is frequently triggered, The `compact gc` means PageStorage need read all of the valid data from the disk, then rewrite it into a new file. Each round of GC brings additional read and write overhead.
+
+Besides these three problems, we also changed the `lock` implements and the `CRC` implements.
+
+![tiflash-ps-v3-architecture](./images/tiflash-ps-v3-architecture.png)
+
+The V3 version of PageStorage consists of two main components, one is `WALStore` and the other is `BlobStore`.
```

Comment:
```suggestion

### Background
We found some problems with the V2 version in the actual production environment.

1. **There is a risk of data loss in the meta part**. As long as the `checksum` and `buffer size` fields in a single meta buffer are damaged at the same time, the subsequent buffers will be unavailable. This situation may happen when the disk failure happens or the meta part is changed by an unexpected operation.
2. **The snapshot of MVCC needs to be optimized**. First, the CPU usage occupied by the snapshot should be reduced. Secondly, the implementation of MVCC structure is too complex to be understood and maintained.
3. **The GC write amplification in the data part is high and the GC task is too heavy**. Since the data part is composed of append write, `compact gc` is frequently triggered. Besides, the meta part is bound to a data part by "PageFile", meaning that we have to apply `compact gc` to compact the small data in the meta part. The `compact gc` means PageStorage need read all of the valid data from the disk, then rewrite it into a new file. Each round of GC brings additional read and write overhead.

Besides these three problems, we also changed the `lock` implements and the `CRC` implements.

### V3 version

![tiflash-ps-v3-architecture](./images/tiflash-ps-v3-architecture.png)

The V3 version of PageStorage consists of two main components, one is `WALStore` and the other is `BlobStore`.
```